sapicore/utils/io.py

```diff
@@ -0,0 +1,270 @@
+            # increment iteration counter.
+            self.iterations += 1
+
+            # add current outputs to this data accumulator instance's cache dictionary, whose values are tensors.
+            for attr in self.attributes:
+                if attr not in self.cache.keys():
+                    # the cache dictionary is empty because this is the first iteration post-dump or in general.
+                    self.cache[attr] = output[attr][None, :]  # expand output by one dimension (zero axis) to fit.
+                else:
+                    # vertically stack output attribute to cache tensor at the appropriate key.
+                    self.cache[attr] = torch.vstack([self.cache[attr], output[attr][None, :]])
+
+            # calculate cache size (MB).
+            cache_size = sum([v.element_size() * v.nelement() for _, v in self.cache.items()]) / 1000000.0
+
+            # if memory usage exceeds the limit given or reached the end of the run, append cached content to HDF5.
+            if cache_size > CHUNK_SIZE or self.iterations == self.entries:
+                with hdf.File(os.path.join(self.log_dir, self.component.identifier + ".h5"), "a") as hf:
+                    for attr in self.attributes:
+                        # increase data chunk size by number of rows in the cache tensor.
+                        self.chunk_size[attr] += self.cache[attr].shape[0]
+                        hf[attr].resize(self.chunk_size[attr], axis=0)
+
+                        # write current cache tensor to its respective HDF5 data structure at the correct indices.
+                        start_index = hf[attr].shape[0] - self.cache[attr].shape[0]
+                        end_index = hf[attr].shape[0]
+
+                        # detach().numpy() necessary if the tensor requires grad.
+                        hf[attr][start_index:end_index, :] = self.cache[attr].cpu().detach().numpy()
+
+                    # clear cache.
+                    self.cache = {}
+
+        return fn
+
+    def forward(self, data: Tensor) -> None:
+        _ = self.component(data)
+        return
+
+
+def log_settings(
+    level: int = logging.INFO,
+    stream: _io.TextIOWrapper = sys.stdout,
+    file: str = None,
+    formatting: str = "%(asctime)s [%(levelname)s] %(message)s",
+):
+    """Temporary basic logger configuration.
+
+    Parameters
+    ----------
+    level: int
+        Logger level, `INFO` by default.
+
+    stream:
+        Destination stream, `stdout` by default.
+
+    file:
+        Destination file, `None` by default.
+
+    formatting: str
+        Message formatting.
+
+    """
+    logging.basicConfig(
+        level=level,
+        format=formatting,
+        handlers=[logging.StreamHandler(stream)] + ([logging.FileHandler(file)] if file else []),
+    )
+
+
+def load_apply_config(configuration: str, apply_to: object = None) -> dict:
```

Comment:
Small thing, but naming ti as `configuration` instead of e.g. `filename` confuses me every time I look at the this. Even at `Pipeline` it confused me. So we should maybe consider a clearer name. E.g. for `Pipeline` I'd even go with something like `config_or_path`, which isn't that weird.

Reply:
In this case it should definitely be changed (only accepts `str`). As for Pipeline, makes sense.

Reply:
Applied your suggestions to `load_apply_config` and `Pipeline`.